### ISSUE TITLE
Start using Scala

### ISSUE THREAD

**Issue by carymrobbins** [carymrobbins/intellij-haskforce#97]
I'll put this out there to see if there are any opinions.

I personally am enjoying Scala and think that it would be a great fit for implementing HaskForce.  Here's a quick pros and cons list off the top of my head -

**Pros**
* Correctness is more easily attainable
* Logic can be more concise and possibly simpler
* Traditional design patterns can still be applied and remain useful.
* Functional idioms are actually possible without depending Java 8.

**Cons**
* Complicates the automated build process, which is important for for CI.
* Potentially more complex code (more language features to be abused).
* May scare away potential contributors.

While these are completely speculative (it might actually *encourage* contributions) they are just a start to get the wheels turning on this.  I think this has the potential for making some things really simple.  However, my biggest concern is the first bullet under the cons section - the build process.  I'm not really sure how complex it will be to get our build moved over to SBT.  Also, this means that the time invested in setting up the project for Scala will take away time from new feature development/bug fixes.  So far it seems like getting stuff done is the most important part (isn't it always?).

Basically, looking for any opinions at all on the topic.  Particularly anyone who has done something similar.  My main goal is to have a strong product that works.

**Comment by carymrobbins** [comment]
HaskForce can now be written in Scala.  However, the majority of the plugin is still written in Java.  As we refactor things we can start converting modules to Scala.  IntelliJ does provide a **Convert to Scala** (Ctrl+G) refactoring option, but it is far from perfect.  Simple modifications to existing Java code should probably just stick to Java.

I experimented with the idea before moving forward, and it seems to provide some real gains.
* Code is far more expressive, concise, composable, and less boilerplate-ridden.
* Pattern matching and exhaustiveness checking is awesome.

To address some of the concerns about using Scala -
* Compile times have been fine for me, I haven't noticed any significant difference from building it purely with Java.
* The IDE support is clearly not as good as Java, but it's still very good.

The biggest problem with using Scala is interop with Java null.  Scala uses `Option[A]` to avoid null, but you still have to deal with it from the IntelliJ platform.  Luckily, Scala's `Option` provides a facility to convert a nullable Java value into a Scala `Option`.

One of the biggest problems with the plugin is the parser.  It would be nice to write a proper parser without Grammar-Kit, and that seems far more feasible with something like Scala as opposed to Java.

We will **not** be using SBT.  I originally tried to replace the current build with it, but as expected, things were way too complicated.  Modifying the existing Ant build was trivial.  However, we are bundling quite a few jars now in the `lib/` dir, so integrating something like Ivy needs to happen pretty soon.
